agent/agentcontainers/api.go

```diff
@@ -263,10 +240,102 @@ func (api *API) loop() {
 	}
 }
 
+// updaterLoop is responsible for periodically updating the container
+// list and handling manual refresh requests.
+func (api *API) updaterLoop() {
+	defer close(api.updaterDone)
+	defer api.logger.Debug(api.ctx, "updater loop stopped")
+	api.logger.Debug(api.ctx, "updater loop started")
+
+	// Ensure that only once instance of the updateContainers is running
+	// at a time. This is a workaround since quartz.Ticker does not
+	// allow us to know if the routine has completed.
+	sema := make(chan struct{}, 1)
+	sema <- struct{}{}
+
+	// Ensure only one updateContainers is running at a time, others are
+	// queued.
+	doUpdate := func() error {
+		select {
+		case <-api.ctx.Done():
+			return api.ctx.Err()
+		case <-sema:
+		}
+		defer func() { sema <- struct{}{} }()
+
+		return api.updateContainers(api.ctx)
+	}
+
+	api.logger.Debug(api.ctx, "performing initial containers update")
+	if err := doUpdate(); err != nil {
+		api.logger.Error(api.ctx, "initial containers update failed", slog.Error(err))
+	} else {
+		api.logger.Debug(api.ctx, "initial containers update complete")
+	}
+	// Signal that the initial update attempt (successful or not) is done.
+	// Other services can wait on this if they need the first data to be available.
+	close(api.initialUpdateDone)
+
+	// Use a ticker func to ensure that doUpdate has run to completion
+	// when advancing time.
+	waiter := api.clock.TickerFunc(api.ctx, api.updateInterval, func() error {
+		err := doUpdate()
```

Comment:
Would it make more sense to have both the periodic update ticker and the refresh trigger both push to a single channel that signals `doUpdate()` to run? That way you could be fairly sure that only one instance of `doUpdate()` is running at a time. This works too, but I'm just more curious about the decision here.